### PR TITLE
Implement ReferenceFree for FinalizeUnchecked

### PR DESCRIPTION
**Pull request by jacob-hughes**

### CODE REVIEW

File: library/core/src/gc.rs [hunk 1]
```diff
@@ -90,9 +97,5 @@ impl<T: ?Sized> DerefMut for FinalizeUnchecked<T> {
...
 unsafe impl<T> FinalizerSafe for FinalizeUnchecked<T> {}
 
-#[unstable(feature = "gc", issue = "none")]
-#[cfg_attr(not(test), rustc_diagnostic_item = "ReferenceFree")]
-pub auto trait ReferenceFree {}
-
-impl<T> !ReferenceFree for &T {}
-impl<T> !ReferenceFree for &mut T {}
+#[cfg(not(bootstrap))]
+unsafe impl<T> ReferenceFree for FinalizeUnchecked<T> {}
```
Comment (ltratt) [inline library/core/src/gc.rs:101]: Can you explain why this is the right thing to do? I don't know that it's the *wrong* thing to do, but I'm not sure either way!

Comment (jacob-hughes) [reply library/core/src/gc.rs:101]: Sure! So `FinalizeUnchecked<T>` is our _escape hatch_ struct. It is unsafe to construct one so users should only be using it if they know what they're doing anyway. The purpose of `FinalizeUnchecked` is to inform FSA that it is being too conservative and that `T`'s finalizer is in fact safe. Without implementing `ReferenceFree`, `FinalizeUnchecked<T>` types will still cause compile-time errors when `T` contains a reference -- even if that reference is never used.

Comment (ltratt) [reply library/core/src/gc.rs:101]: OK, I can buy that.

**Comment by jacob-hughes** [comment]
Fixed here 01acc4a986d0123cef586376b0c9323ccef6cb53

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**